✨  Add ClusterClass patch engine

sbueringer opened the pull request:
<!-- please add a icon to the title of this PR (see https://sigs.k8s.io/cluster-api/VERSIONING.md), and delete this line and similar ones -->
<!-- the icon will be either ⚠️ (:warning:, major or breaking changes), ✨ (:sparkles:, feature additions), 🐛 (:bug:, patch and bugfixes), 📖 (:book:, documentation or proposals), or 🌱 (:seedling:, minor or other) -->

**What this PR does / why we need it**:
This PR implements the patch engine and variable calculation described in "Applying Patches" in: https://github.com/kubernetes-sigs/cluster-api/issues/5460 (Note: there are also code walkthroughs linked in the issue, the second one ~ still matches the code from this PR).

Goal of this PR is to add the patch engine and variable calculation. 

The implementation for the inline JSON patches will be added after this PR has been merged via https://github.com/kubernetes-sigs/cluster-api/pull/5583.


**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

ykakarap commented:
This is great @sbueringer!! 🚀 


First pass at a high level review.
The overall implementation seems good. 👍

Some high level suggestions: 
- It would be nice if we can make a core a little more readable. Some of the variable and type names do not reflect their purpose/intent. Some examples: 
   - `GenerateRequest` - >`Request`? (similarly with `GenerateResponse`)
   -  `requestToDesiredState` -> `updateDesiredState`
   - The variable `patches`/`patch` is used in several places with different meaning. In some places it is the `GenerateResponse`, in some places it is the `Patch` from the cluster class, etc. Can we rename some of the places to make the variable naming more consistent?
- Breaking down `engine.go` into several files. Grouping some of the patch functions (like `patchObject`, `patchTemplate`, `patchTemplateSpec`), the helper functions (like `copySpec`, `bytesToUnstructured`), etc into their own files would help with code maintenance.


Leaving this at a high level review for now as this PR is still a WIP (there are also a few TODOs in the code).

Will do a full review once PR is ready. 🙂

sbueringer commented:
> * It would be nice if we can make a core a little more readable. Some of the variable and type names do not reflect their purpose/intent. Some examples:
>   
>   * `GenerateRequest` - >`Request`? (similarly with `GenerateResponse`)
>   * `requestToDesiredState` -> `updateDesiredState`
>   * The variable `patches`/`patch` is used in several places with different meaning. In some places it is the `GenerateResponse`, in some places it is the `Patch` from the cluster class, etc. Can we rename some of the places to make the variable naming more consistent?
> * Breaking down `engine.go` into several files. Grouping some of the patch functions (like `patchObject`, `patchTemplate`, `patchTemplateSpec`), the helper functions (like `copySpec`, `bytesToUnstructured`), etc into their own files would help with code maintenance.

Thx, so far! :)

Some answers:
* GenerateRequest and GenerateResponse were named like this as it's a common pattern in Kubernetes and gRPC to include the operation in the Request/Response name, e.g. AdmissionRequest, AdmissionResponse, TokenRequest, CertificateSigningRequest. I think it's about namespacing in general. Note: right now those structs are only part of an internal interface
* `requestToDesiredState` -> `updateDesiredState` > Done
* I'll rename Request/Response variables to req/resp where necessary
* I'll move some helpers to patch.go.  wouldn't split it up further as it doesn't necessarily make it more maintainable if the code is distributed over multiple files. (especially as the only consumer right now is the engine)

randomvariable commented:
I probably won't be able to get round to a full review before Thursday due to other stuff going on, will queue up.

sbueringer commented on controllers/topology/internal/extensions/patches/engine_test.go:
```diff
@@ -0,0 +1,542 @@
+	. "sigs.k8s.io/cluster-api/internal/matchers"
+)
+
+func TestApply(t *testing.T) {
```
I've currently only added one top-level unit test for `engine.go`. It seems way simpler to just have a few more cases on that level compared to more or less duplicating the whole test setup for `request`, `applyPatchesToRequest` and `updateDesiredState`. Please let's discuss if there are other opinions on that :)

sbueringer commented on controllers/topology/internal/extensions/patches/patch_test.go:
```diff
@@ -0,0 +1,369 @@
+	"sigs.k8s.io/cluster-api/controllers/topology/internal/contract"
+)
+
+func TestCopySpec(t *testing.T) {
```
Similar to engine.go, I've only added one test for copySpec. If preferred I can duplicate the test cases for the 3 patch funcs. 

I thought testing copySpec covers the most cases and the 3 other funcs are transitively tested by the tests in engine_test.go

sbueringer commented:
PR should be ready for review now. (open TODOs are supposed to be solved in other PRs)

(cc @ykakarap)

fabriziopandini commented on controllers/topology/internal/extensions/patches/patch.go:
```diff
@@ -0,0 +1,198 @@
+		destSpecPath:     "spec",
+		fieldsToPreserve: patchOptions.preserveFields,
+	}); err != nil {
+		return errors.Wrapf(err, "failed to patch object %s by appying changes from the patched template",
```
```suggestion
		return errors.Wrapf(err, "failed to apply patch to %s",
```

fabriziopandini commented on controllers/topology/internal/extensions/patches/patch.go:
```diff
@@ -0,0 +1,198 @@
+	}
+
+	// Log the delta between the object before and after applying the accumulated patches.
+	helper, err := mergepatch.NewHelper(original, object, nil)
```
I'm not sure it is appropriate using mergepatch.NewHelper to track the difference between objects, given that it computes a two way patch which preserve values from the original object, while instead we are overriding the entire spec (same for patch template)

sbueringer replied:
100% agree, it's not appropriate. I'll add a small diff util func

sbueringer replied:
Fixed, PTAL

fabriziopandini commented on controllers/topology/internal/extensions/patches/patch.go:
```diff
@@ -0,0 +1,198 @@
+
+// ApplyOptions applies the given patch options
+// and then returns itself (for convenient chaining).
+func (o *PatchOptions) ApplyOptions(opts []PatchOption) *PatchOptions {
```
nit
```suggestion
func (o *PatchOptions) ApplyOptions(opts []PatchOption) {
```

fabriziopandini commented on controllers/topology/internal/extensions/patches/patch.go:
```diff
@@ -0,0 +1,198 @@
+}
+
+// patchTemplateSpec overwrites spec in templateJSON with spec of patchedTemplateBytes.
+func patchTemplateSpec(templateJSON *apiextensionsv1.JSON, patchedTemplateBytes []byte) error {
```
Wondering if we can drop this func and use patchTemplate.
They seems functionally equivalent (the only part of spec we care about spec.template.spec), so the only problem to be addressed is the difference of type, so we need a wrapper, something like patchRequestTemplate

sbueringer replied:
It's more than that:
* patchTemplate patches: .spec.template.spec => spec.template.spec. 
* patchTemplateSpec patches: .spec => .spec

Except we want to change the semantic and only want to allow patches to modify `.spec.template.spec` instead of the entire `.spec`. The only other currently existing .spec  field in templates I can think of is `.spec.template.metadata`.

fabriziopandini replied:
> The only other currently existing .spec field in templates I can think of is .spec.template.metadata.

yep, that's my point as well. this is why I think we should dedup

sbueringer replied:
The 99% use case is definitely `.spec.template.spec`. We can also start with that and consider expanding on that it if there is demand

sbueringer replied:
I refactored patchTemplate/patchObject a bit to deduplicate code and to have a correct diff. I don't think we can reuse patchTemplate or patchUnstructured here as we don't need the same diff logic and logging.

Independent of that we can of course still decide to reduce the scope of potential changes to .spec.template.spec

fabriziopandini commented on controllers/topology/internal/extensions/patches/engine.go:
```diff
@@ -0,0 +1,495 @@
+// Apply applies patches to the desired state according to the patches from the ClusterClass, variables from the Cluster
+// and builtin variables.
+// * A GenerateRequest with all templates and global and template-specific variables is created.
+// * Then for all patches of a ClusterClass, patches are generated and successively applied to the templates
```
```suggestion
// * Then for all patches definitions of a ClusterClass, patches are generated and successively applied to the templates
```

sbueringer replied:
I've went with a slightly different wording to deduplicate "patches"

fabriziopandini commented on controllers/topology/internal/extensions/patches/engine.go:
```diff
@@ -0,0 +1,495 @@
+		return errors.Wrapf(err, "failed to generate patch request")
+	}
+
+	// Loop over patches in ClusterClass, generate patches and apply them to the request.
```
```suggestion
	// Loop over patches in ClusterClass, generate patches and apply them to the request 
	// respecting the order they are defined.
```

fabriziopandini commented on controllers/topology/internal/extensions/patches/engine.go:
```diff
@@ -0,0 +1,495 @@
+
+	// Loop over patches in ClusterClass, generate patches and apply them to the request.
+	for i := range blueprint.ClusterClass.Spec.Patches {
+		patch := blueprint.ClusterClass.Spec.Patches[i]
```
```suggestion
		patchDefinition := blueprint.ClusterClass.Spec.Patches[i]
```
To disambiguate from the generated patch

sbueringer replied:
Not sure if we have to disambiguate as the generated patch is only used inside of `applyPatchesToRequest` and I think it's better to keep the var names short versus trying to keep the var names unique per struct.

(I would have to name the var `clusterClassPatch` as that's the corresponding struct)

fabriziopandini replied:
Not blocking, but as it is not it read that we have a patch and then we use it to generated other patches, and this could be confusing at a first sight. this is why I suggested a more verbose name

sbueringer replied:
Okay that makes sense. I thought it's about using different variable names for different structs in general as brought up here https://github.com/kubernetes-sigs/cluster-api/pull/5534#issuecomment-957030201

sbueringer replied:
I'll call it `clusterClassPatch` though to avoid the impression that it's a `PatchDefinition`.

fabriziopandini commented on controllers/topology/internal/extensions/patches/engine.go:
```diff
@@ -0,0 +1,495 @@
+			return errors.Errorf("failed to generate patches for patch %q", patch.Name)
+		}
+
+		// Apply patches to the request.
```
```suggestion
		// Apply patches to the request.
		// NOTE: All the partial patches accumulate on top of the request, so the following
		// patch generator will get the modified version if templates.
```

fabriziopandini commented on controllers/topology/internal/extensions/patches/engine.go:
```diff
@@ -0,0 +1,495 @@
+	}
+
+	// Use patched templates to update the desired state objects.
+	log.V(5).Infof("Applying patches to desired state")
```
```suggestion
	log.V(5).Infof("Applying patched templates to desired state")
```

fabriziopandini commented on controllers/topology/internal/extensions/patches/engine.go:
```diff
@@ -0,0 +1,495 @@
+// * Then for all patches of a ClusterClass, patches are generated and successively applied to the templates
+//   in the GenerateRequest.
+// * Eventually the patched templates are used to update the specs of the desired objects.
+func Apply(ctx context.Context, blueprint *scope.ClusterBlueprint, desired *scope.ClusterState) error {
```
Let's return early if there are no patch definitions

sbueringer replied:
Good point

sbueringer commented:
Rebase on top of main, delta:
* Aligned to changes in the MachineDeploymentClass builder
* Fixed the TODO about requiring `IgnoreAutogeneratedMetadata` because the EqualObject helper is fixed now
* Changed enum in the api package to use `iota`

fabriziopandini commented on controllers/topology/internal/extensions/patches/engine.go:
```diff
@@ -0,0 +1,495 @@
+		TargetType: api.InfrastructureClusterTemplateTargetType,
+	})
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to create InfrastructureCluster template %s",
```
```suggestion
		return nil, errors.Wrapf(err, "failed to prepare InfrastructureCluster template %s for patching",
```
I would avoid using create given that everything happens in memory
Same for other error message in this func

fabriziopandini commented on controllers/topology/internal/extensions/patches/engine.go:
```diff
@@ -0,0 +1,495 @@
+
+	// Add BootstrapConfigTemplate and InfrastructureMachine template for all MachineDeploymentTopologies
+	// in the Cluster.
+	// NOTE: We intentionally iterate over MachineDeploymentTopologies in the Cluster instead of over
+	// MachineDeploymentClasses in the ClusterClass. We have to calculate patches per MachineDeploymentTopology
+	// instead of per MachineDeploymentClass, because the builtin variable values depend on the current
+	// MachineDeployment. E.g., there are variables for the version or replicas of the MachineDeployment.
```
```suggestion
	// NOTE: We intentionally iterate over MachineDeployment in the Cluster instead of over
	// MachineDeploymentClasses in the ClusterClass because each MachineDeployment in a topology
	// has its own state, e.g. version or replicas, and it is required to use those info as built in variable 
	// e.g. to compute the machine image for a specific Kubernetes version.
```

fabriziopandini commented on controllers/topology/internal/extensions/patches/engine.go:
```diff
@@ -0,0 +1,495 @@
+	// instead of per MachineDeploymentClass, because the builtin variable values depend on the current
+	// MachineDeployment. E.g., there are variables for the version or replicas of the MachineDeployment.
+	for mdTopologyName, md := range desired.MachineDeployments {
+		// Lookup MachineDeploymentTopology.
```
```suggestion
		// Lookup MachineDeploymentTopology definition from Cluster.Topology.
```
```suggestion
		// Lookup MachineDeploymentTopology.
```

fabriziopandini commented on controllers/topology/internal/extensions/patches/engine.go:
```diff
@@ -0,0 +1,495 @@
+			return nil, err
+		}
+
+		// Get corresponding MachineDeploymentClass.
```
```suggestion
		// Get corresponding MachineDeploymentClass from the ClusterClass.
```

fabriziopandini commented on controllers/topology/internal/extensions/patches/engine.go:
```diff
@@ -0,0 +1,495 @@
+			return &mdTopology, nil
+		}
+	}
+	return nil, errors.Errorf("failed to lookup MachineDeployment topology %q in Cluster topology", mdTopologyName)
```
```suggestion
	return nil, errors.Errorf("failed to lookup MachineDeployment topology %q in Cluster.spec.topology.workers", mdTopologyName)
```

fabriziopandini commented on controllers/topology/internal/extensions/patches/engine.go:
```diff
@@ -0,0 +1,495 @@
+var createPatchGenerator = createPatchGeneratorImpl
+
+// createPatchGeneratorImpl creates a patch generator for the given patch.
+// NOTE: Currently only inline JSON patches are supported and it's not clear yet how
+// e.g. external patches will be configured in the future.
+// Note: This func is required to be able to inject a fake patch generator for unit tests.
```
```suggestion
// NOTE: Currently only inline JSON patches are supported; in the future we will add
// external patches as well.
```

fabriziopandini commented on controllers/topology/internal/extensions/patches/engine.go:
```diff
@@ -0,0 +1,495 @@
+
+		// If a patch doesn't apply to any template, this is a misconfiguration.
+		if template == nil {
+			return errors.Errorf("could not find template with id %q in GenerateRequest", templateIDRef(patch.TemplateID))
```
```suggestion
			return errors.Errorf("generated patch is targeted a template with id %q that does not exists", templateIDRef(patch.TemplateID))
```

fabriziopandini commented on controllers/topology/internal/extensions/patches/engine.go:
```diff
@@ -0,0 +1,495 @@
+
+// updateDesiredState uses the patched templates of a GenerateRequest to update the desired state.
+// NOTE: This func should be called after all the patches have been applied to the GenerateRequest.
+// TODO(sbueringer): Let's only overwrite desired state objects if the corresponding template actually have been patched.
```
follow up PR?

sbueringer replied:
Given that it's an optimization and I'm not sure if the current PR is already to big, I would say yes.

sbueringer replied:
Looks like I can fix it when implementing the correct diff.

fabriziopandini reviewed (commented):
@sbueringer great stuff!
Mostly nits; the only thing that requires some deep dive is the usage of mergepatch.
I still need to take a better look at test for the engine

sbueringer commented on docs/proposals/202105256-cluster-class-and-managed-topologies.md:
```diff
@@ -575,7 +575,7 @@ It’s also possible to use so-called builtin variables in addition to user-defi
 - `cluster.topology.controlPlane.{labels,annotations,replicas,version}`
     - **Note**: these variables are only available when patching control plane or control plane machine templates.
-- `cluster.topology.machineDeployment.current.{labels,annotations,replicas,version,class}`
+- `cluster.topology.machineDeployment.current.{labels,annotations,replicas,version,class,name,topologyName}`
```
My plan was to drop it in a separate PR depending on the feedback (or lack thereof) on https://github.com/kubernetes-sigs/cluster-api/issues/5577. Mainly to avoid controversial discussions about it on this PR.

sbueringer commented on controllers/topology/internal/extensions/patches/variables/variables.go:
```diff
@@ -0,0 +1,158 @@
+
+var (
+	// Builtin is the prefix of all builtin variables.
+	Builtin = "builtin"
```
Done, although I can only do it for this one, everything else has to be a variable because of the function call.

sbueringer commented on controllers/topology/internal/extensions/patches/variables/variables.go:
```diff
@@ -0,0 +1,158 @@
+
+	// Builtin ControlPlane variables.
+
+	// BuiltinClusterTopologyControlPlaneReplicas is the value of the replicas field of the ControlPlane.
```
I rephrased it a bit, but I'm not sure your suggestion is true. Actually I'm taking those values from the desired objects. Mainly that the values which will be in the desired objects after the reconcile can be used. When I take them from the topology the version especially will often be the wrong one.

fabriziopandini replied:
makes sense, you can ignore this suggestion

sbueringer commented on controllers/topology/internal/extensions/patches/variables/variables.go:
```diff
@@ -0,0 +1,158 @@
+	// BuiltinClusterTopologyMachineDeploymentClass is the class name of the MachineDeployment.
+	BuiltinClusterTopologyMachineDeploymentClass = builtinVariable("cluster.topology.machineDeployment.current.class")
+	// BuiltinClusterTopologyMachineDeploymentName is the name of the MachineDeployment.
+	BuiltinClusterTopologyMachineDeploymentName = builtinVariable("cluster.topology.machineDeployment.current.name")
```
I'll update based on the outcome of https://github.com/kubernetes-sigs/cluster-api/pull/5534#discussion_r742943974.

I agree the current semantic is that we calculate the values based on the desired objects so we should drop topology for all ControlPlane/MD variables.

fabriziopandini replied:
Agreed. Let's drop topology from the name of all the variables that are derived from the computed objects

sbueringer replied:
I would be fine with it, I think it was a suggestion from @vincepri to add it. Let's see

sbueringer replied:
New version will be:

- `builtin.cluster.{name,namespace,labels,annotations}`
- `builtin.cluster.topology.{version,class}`
- `builtin.controlPlane.{labels,annotations,replicas,version}`
    - **Note**: these variables are only available when patching control plane or control plane machine templates.
- `builtin.machineDeployment.{labels,annotations,replicas,version,class,name,topologyName}`
    - **Note**: these variables are only available when patching MachineDeployment templates and contain the values 
      of the current `MachineDeploymentTopology`.

sbueringer commented on controllers/topology/internal/extensions/patches/variables/variables.go:
```diff
@@ -0,0 +1,158 @@
+func ControlPlane(controlPlaneTopology *clusterv1.ControlPlaneTopology, controlPlane *unstructured.Unstructured) (VariableMap, error) {
+	variables := VariableMap{}
+
+	if controlPlaneTopology.Replicas != nil {
```
I've rephrased a bit, given that the field doesn't have to be set even with KCP when the default value should be used, i.e. the topology reconciler just doesn't reconcile the replicas field.

sbueringer commented on controllers/topology/internal/extensions/patches/api/interface.go:
```diff
@@ -0,0 +1,149 @@
+)
+
+// Generator defines a component that can generate patches for ClusterClass templates.
+// NOTE: We are introducing this interface as a decoupling layer between the patch engine and the concrete components
```
Good point, done

sbueringer commented on controllers/topology/internal/extensions/patches/api/interface.go:
```diff
@@ -0,0 +1,149 @@
+	Kind string
+
+	// TargetType defines where the template is used.
+	TargetType TargetType
```
Sounds good. +1 for the new name and for avoiding introducing the "target" concept.

I'll wait a bit for further feedback from others.

sbueringer replied:
I'll change it to TemplateType

sbueringer commented on controllers/topology/internal/extensions/patches/patch.go:
```diff
@@ -0,0 +1,198 @@
+	return nil
+}
+
+// patchTemplate overwrites spec.template.spec in template with spec.template.spec of patchedTemplate,
+// while preserving the configured fields.
```
I added an hypothetical example because otherwise it would be misleading as I'm currently never preserving fields when patching templates.

fabriziopandini commented on controllers/topology/internal/extensions/patches/engine_test.go:
```diff
@@ -0,0 +1,541 @@
+			},
+		},
+		{
+			name: "Should apply JSON patches to MachineDeployment templates",
```
Q: should we have a patch that apply to all machineDeployments?

sbueringer replied:
I added a test case to patch the InfraMachineTemplates of both MDs with the same patch (via two individual patches).

Did you mean that or did you mean I should use a patch which has TemplateType "MachineDeployment/InfrastructureMachineTemplate" and has no MachineDeployment.TopologyName and MachineDeployment.Class set and should thus apply to both?
(That's currently not supported because currently the templateID of the response must be equal to the one of the request and we're always setting those two fields for MD templates)

fabriziopandini replied:
I was thinking to a patch without MachineDeployment.TopologyName and MachineDeployment.Class, thus avoiding the users to define via two or more individual patches to apply the same change.
But I'm also fine to open an issue and address this improvement down the line

sbueringer replied:
In this case those fields are filled on the Request by the Patch Engine on the Response by the json patch generator. (or later on potentially by external patches)

The selector on the patch struct which the user fills out requires to specify MachineDeployment names to match MachineDeployments, but we wanted to have that explicit.

I think it would be good to keep it strict for now and later on when we define the patch extension API we should nail down if we want to support matching all MDs in that case or only explicitly matched ones (explicit matching would be inline with how the selector in the ClusterClass behaves)

fabriziopandini commented on controllers/topology/internal/extensions/patches/engine_test.go:
```diff
@@ -0,0 +1,541 @@
+				expectedBootstrapTemplates[mdTopology] = md.BootstrapTemplate.DeepCopy()
+				expectedInfrastructureMachineTemplate[mdTopology] = md.InfrastructureMachineTemplate.DeepCopy()
+			}
+			// Set expected fields.
```
```suggestion
			// Set expected fields on the copy of the objects, so they can be used for comparison with the Apply result.
```

fabriziopandini commented on controllers/topology/internal/extensions/patches/engine_test.go:
```diff
@@ -0,0 +1,541 @@
+		t.Run(tt.name, func(t *testing.T) {
+			g := NewWithT(t)
+
+			// Set up test objects.
```
Let's add a brief comment about what the test objects are

fabriziopandini commented on controllers/topology/internal/extensions/patches/engine_test.go:
```diff
@@ -0,0 +1,541 @@
+	return blueprint, desired
+}
+
+type fakePatchGenerator struct {
```
Lets add a comment about what the fakePatchGenerator is (a patch generator that returns the patches provided in input)

sbueringer commented on controllers/topology/internal/extensions/patches/api/interface.go:
```diff
@@ -0,0 +1,149 @@
+// PatchType define the type for patch types enum.
+type PatchType int32
+
+const (
+	// JSONPatchType identifies a https://datatracker.ietf.org/doc/html/rfc6902 json patch.
+	JSONPatchType PatchType = 0
+
+	// JSONMergePatchType identifies a https://datatracker.ietf.org/doc/html/rfc7386 json merge patch.
+	JSONMergePatchType PatchType = 1
+)
```
Nope, sounds good, just did the same for the other enum in a previous commit

sbueringer replied:
Fixed

sbueringer commented on controllers/topology/internal/extensions/patches/util.go:
```diff
@@ -0,0 +1,165 @@
+/*
```
Good point, thx. Done

vincepri commented on controllers/topology/desired_state.go:
```diff
@@ -29,14 +29,17 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
 )
 
+var patchEngine = patches.NewEngine()
+
```
Could this be passed to the reconciler instead?

sbueringer replied:
Sure, just to confirm that I got it right:
* add a patchEngine field in the `ClusterReconciler`
* set it to `patches.NewEngine()` in `controllers/topology.SetupWithManager`
?

vincepri replied:
Correct

sbueringer replied:
Sounds good. It was a bit awkward to just use a package variable :/

Done

vincepri commented on controllers/topology/internal/extensions/patches/api/interface.go:
```diff
@@ -0,0 +1,138 @@
+	Generate(ctx context.Context, request *GenerateRequest) (*GenerateResponse, error)
+}
+
+// GenerateRequest defines the input for a Generate request.
+type GenerateRequest struct {
+	// Variables is a name/value map containing variables.
+	Variables map[string]apiextensionsv1.JSON
+
+	// Items contains the list of templates to generate patches for.
+	Items []*GenerateRequestTemplate
+}
```
Should we add json tags to each one of these structs?

sbueringer replied:
I can do it. I'm not sure if we should already, given that we currently don't marshal/unmarshal those structs and we don't know if we need JSON tags, gRPC tags, both or maybe even define the structs in proto.

vincepri replied:
That's fair, let's do it later

fabriziopandini commented:
lgtm pending squash. great stuff!

sbueringer commented:
> lgtm pending squash. great stuff!

Done :)

vincepri reviewed (approved):
/lgtm

fabriziopandini commented:
/lgtm
/approve
